gobblin-modules/gobblin-couchbase/src/main/java/gobblin/couchbase/writer/CouchbaseWriter.java

```diff
@@ -103,38 +128,132 @@ public CouchbaseWriter(CouchbaseEnvironment couchbaseEnvironment, Config config)
 
+
+    _operationTimeout = 2000;
+    _operationTimeunit = TimeUnit.MILLISECONDS;
+
   }
 
 
+  @VisibleForTesting
   Bucket getBucket()
   {
     return _bucket;
   }
 
+
   @Override
-  public void cleanup()
-      throws IOException {
+  public Future<WriteResponse> write(final D record, final WriteCallback callback) {
+    if (record instanceof TupleDocument) {
+      ((TupleDocument) record).content().value1().retain();
+    }
+    Observable<D> observable = _bucket.async().upsert(record);
+    if (callback == null) {
+      return new WriteResponseFuture<>(observable.timeout(_operationTimeout, _operationTimeunit)
+          .toBlocking().toFuture(), new GenericWriteResponseWrapper<D>());
+    } else {
+
+      final AtomicBoolean callbackFired = new AtomicBoolean(false);
+      final BlockingQueue<Pair<WriteResponse, Throwable>> writeResponseQueue = new ArrayBlockingQueue<>(1);
+
+      final Future<WriteResponse> writeResponseFuture = new Future<WriteResponse>() {
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+          return false;
+        }
+
+        @Override
+        public boolean isCancelled() {
+          return false;
+        }
+
+        @Override
+        public boolean isDone() {
+          return callbackFired.get();
+        }
+
+        @Override
+        public WriteResponse get()
+            throws InterruptedException, ExecutionException {
+          Pair<WriteResponse, Throwable> writeResponseThrowablePair = writeResponseQueue.take();
+          return getWriteResponseorThrow(writeResponseThrowablePair);
+        }
 
+        @Override
+        public WriteResponse get(long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+          Pair<WriteResponse, Throwable> writeResponseThrowablePair = writeResponseQueue.poll(timeout, unit);
+          if (writeResponseThrowablePair == null) {
+            throw new TimeoutException("Timeout exceeded while waiting for future to be done");
+          } else {
+            return getWriteResponseorThrow(writeResponseThrowablePair);
+          }
+        }
+      };
+
+      observable.timeout(_operationTimeout, _operationTimeunit).subscribe(new Subscriber<D>() {
+        @Override
+        public void onCompleted() {
+        }
+
+        @Override
+        public void onError(Throwable e) {
+          callbackFired.set(true);
+          writeResponseQueue.add(new Pair<WriteResponse, Throwable>(null, e));
+          callback.onFailure(e);
+        }
+
+        @Override
+        public void onNext(D doc) {
+          callbackFired.set(true);
+          WriteResponse writeResponse = new GenericWriteResponse<D>(doc);
+          writeResponseQueue.add(new Pair<WriteResponse, Throwable>(writeResponse, null));
+          callback.onSuccess(writeResponse);
+          if (doc instanceof TupleDocument) {
```

Comment:
What if doc is not of TupleDocument type? Also, wondering if content().value1().release() method chain can yield potential NPE?